swap/src/database.rs

```diff
@@ -210,34 +236,49 @@ impl Database {
             .context("Could not flush db")
     }
 
-    pub fn get_state<T: DeserializeOwned>(&self, swap_id: Uuid) -> Result<T> {
+    pub fn get_state(&self, swap_id: Uuid) -> Result<T> {
         let key = serialize(&swap_id)?;
 
         let encoded = self
             .swaps
             .get(&key)?
             .ok_or_else(|| anyhow!("Swap with id {} not found in database", swap_id))?;
 
-        let state = deserialize(&encoded).context("Could not deserialize state")?;
-        Ok(state)
+        let state= deserialize(&encoded).context("Could not deserialize state")?;
+
+        let t = T::try_from(state);
+
+        let t = match t {
+            Ok(a) => a,
+            Err(_) => bail!("a")
+        };
+
+        Ok(t)
     }
 
-    pub fn all<T: DeserializeOwned>(&self) -> Result<Vec<(Uuid, T)>> {
+    pub fn all(&self) -> Result<Vec<(Uuid, T)>> {
         self.all_iter().collect()
     }
 
-    fn all_iter<T: DeserializeOwned>(&self) -> impl Iterator<Item = Result<(Uuid, T)>> {
+    fn all_iter(&self) -> impl Iterator<Item = Result<(Uuid, T)>> {
         self.swaps.iter().map(|item| {
             let (key, value) = item.context("Failed to retrieve swap from DB")?;
 
             let swap_id = deserialize::<Uuid>(&key)?;
-            let swap = deserialize::<T>(&value).context("Failed to deserialize swap")?;
+            let swap = deserialize::<Swap>(&value).context("Failed to deserialize swap")?;
+
+            let t = T::try_from(swap);
+
+            let t = match t {
+                Ok(a) => a,
+                Err(_) => bail!("a")
+            };
```

Comment:
That is equivalent to calling `.context("a")` :)

Reply:
If i remember correctly, I tried this but there was a really weird error and I had to this workaround